### PR TITLE
[Maps] Add polylines and polygons

### LINKED ISSUE #1673: [Enhancement] Map with shapes from pin
## Rationale

UX could reasonably want to overly shapes onto maps. For example, a semi-transparent circle centered on a pin.

## Implementation

Could make it complicated and general or could just make it this:

```csharp
public class Pin {
	public static readonly BindableProperty CircleColor; // color
	public static readonly BindableProperty CircleSize; // double
}
```

## Expected Result

A circular map overlay centered at the pin.

## Implications for CSS

None.

## Backward Compatibility

None.

## Difficulty : Easy

### LINKED ISSUE #5773: [Enhancement] Expand data binding for Map control
## Summary
In one of the latest Xamarin Forms updates, we received data binding for `Pins` - [https://docs.microsoft.com/en-us/xamarin/xamarin-forms/user-interface/map#populating-a-map-with-data-using-data-binding](https://docs.microsoft.com/en-us/xamarin/xamarin-forms/user-interface/map#populating-a-map-with-data-using-data-binding).

Would be great if we could also get this for `Polylines`.

## API Changes

Include `Polylines` property on the `Map` control for data binding options, enabling collections to be used for drawing lines on the map.

## Intended Use Case
MVVM all the things

### PR DESCRIPTION
**Pull request by jcmanke**
### Description of Change ###

Add the ability to draw polylines and polygons on a Map.

### Issues Resolved ### 

- fixes #1673
- fixes #5773

### API Changes ###

Added:
 - Maps
   - `public class MapElement : Element`
   - `public class Polyline : MapElement`
   - `public class Polygon : MapElement`
   - `public IList<MapElement> Map.MapElements { get; }`
 - Android MapRenderer
   - `protected APolyline GetNativePolyline(Polyline polyline)`
   - `protected Polyline GetFormsPolyline(APolyline polyline)`
   - `protected virtual PolylineOptions CreatePolylineOptions(Polyline polyline)`
   - `protected APolygon GetNativePolygon(Polygon polygon)`
   - `protected Polygon GetFormsPolygon(APolygon polygon)`
   - `protected virtual PolygonOptions CreatePolygonOptions(Polygon polygon)`
 - iOS MapRenderer
   - `protected virtual MKOverlayRenderer GetViewForOverlay(MKMapView mapview, IMKOverlay overlay)`
   - `protected virtual MKPolylineRenderer GetViewForPolyline(MKPolyline mkPolyline)`
   - `protected virtual MKPolygonRenderer GetViewForPolygon(MKPolygon mkPolygon)`
 - UWP MapRenderer
   - `protected virtual MapPolyline LoadPolyline(Polyline polyline)`
   - `protected virtual MapPolygon LoadPolygon(Polygon polygon)`
   - `protected Geopath PositionsToGeopath(IList<Position> positions)`

Changed:
 - Renamed `OnCollectionChanged` method to `OnPinCollectionChanged` in Android, iOS, and UWP MapRenderers

### Platforms Affected ### 

- Core/XAML (all platforms)
- iOS
- Android
- UWP

### Behavioral/Visual Changes ###

None

### Before/After Screenshots ### 

Screenshot_1564691890

2019-08-01_03-41-24-PM

map_elements_uwp

### Testing Procedure ###

Added Map Elements Gallery page to Control Gallery app. 

### PR Checklist ###

- [ ] Has automated tests 
- [x] Rebased on top of the target branch at time of PR
- [x] Changes adhere to coding standard

### CODE REVIEW

**Comment by jcmanke** [comment]
@andreinitescu I imagine you might have some good input on this.

**Comment by samhouts** [comment]
Maps are getting so much love lately! Thanks! ❤️

**Comment by JKennedy24** [comment]
Is this just porting the code across from this sample?

https://developer.xamarin.com/samples/xamarin-forms/CustomRenderers/Map/Polyline/

**Comment by jcmanke** [comment]
Not directly at least. I'm porting from a custom Map subclass I use at work, but we may used that sample when initially writing our implementation.

File: Xamarin.Forms.Maps/Polyline.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+		public float StrokeWidth
+		{
+			get => (float)GetValue(StrokeWidthProperty);
+			set => SetValue(StrokeWidthProperty, value);
+		}
+
+		public IList<Position> Geopath { get; }
+
+		[EditorBrowsable(EditorBrowsableState.Never)]
```
Comment (hartez) [inline Xamarin.Forms.Maps/Polyline.cs]: Why the `EditorBrowsableState.Never` for `PolylineId`?

Comment (jcmanke) [reply Xamarin.Forms.Maps/Polyline.cs]: It's a analogue to `Pin.MarkerId`, which also has the attribute. Something for the renderers to match the cross-platform object to its native counterpart, shouldn't really be used outside of that.

Comment (hartez) [reply Xamarin.Forms.Maps/Polyline.cs]: Could it be made `internal`?

Comment (jcmanke) [reply Xamarin.Forms.Maps/Polyline.cs]: Only if you're okay with adding an `InternalsVisibleToAttribute` back to the Maps project. This property is defined in the shared project and used in the renderers.

Comment (hartez) [reply Xamarin.Forms.Maps/Polyline.cs]: Argh. Good point. `public` it is, then.

**Comment by jcmanke** [comment]
Any objections to me expanding this PR to include Polygons and Circles as well, instead of adding them piecemeal? I intend to basically follow the API described by @andreinitescu in this comment: https://github.com/xamarin/Xamarin.Forms/issues/1673#issuecomment-438243683

**Comment by jcmanke** [comment]
Added a MapElement base class and Polygons. Circles are a little bit trickier because they are not natively supported in UWP the same way Android and iOS, but can be done by either drawing a MapPolygon that is approximately circular, or using an Ellipse similar to how the user location is presented.

**Comment by jcmanke** [comment]
@hartez @kingces95 Please review the changes I made to also include Polygons.

**Comment by jfversluis** [comment]
@jcmanke there is a build error:

> "Xamarin.Forms.Maps.iOS\MapRenderer.cs(514,23): Error CS0308: The non-generic type 'IEnumerable' cannot be used with type arguments"

Also we need to do a bit of re-reviewing because of the polygon functionality that was added :)

**Comment by matheuscschenfeld** [comment]
How can I add a sector map Tap recognizer? For example, I have a polygon on map and I need estou detect if user tapped this polygon, if its tapped change the FillColor to show that the polygon was selected

**Comment by jcmanke** [comment]
On Android, you can register for the `PolylineClicked`/`PolygonClicked` event(s) on the `GoogleMap`.

On iOS, you would have to add a `UITapGestureRecognizer` to the `MKMapView` and detect if the tap is inside the bounds of the overlay. (See https://stackoverflow.com/questions/48045781/xamarin-ios-map-circle-overlay-click for an example.)

On UWP, register for the `MapElementClick` event on the `MapControl`.

**Comment by matheuscschenfeld** [comment]
Good Morning, I have one issue here:
Contextualizing the situation: 
I have one MKMapView in my CustomMapRender(iOS), in this map I have some polygons and renders of which polygon, according to the image below:
Simulator Screen Shot - iPhone 6 - 2019-12-11 at 19 00 42
As you can see, here one of the polygons is green, its thats why when the user selects one polygon it changes the color, in this way user can see that this polygon is selected. 
to create this polygons I use the code:

```
if (e.NewElement != null)
            {
                var formsMap = (CustomMap)e.NewElement;
                var nativeMap = Control as MKMapView;
                List<App.Graphic.Polygon> Polygons;

                if (nativeMap != null && formsMap != null)
                {
                    try
                    {
                        Polygons = formsMap.Polygons;
                        nativeMap.OverlayRenderer = GetOverlayRenderer;
                        for (int i = 0; i < Polygons.Count; i++)
                        {
                            CLLocationCoordinate2D[] coords = new CLLocationCoordinate2D[Polygons[i].Vertices.Count];
                            int index = 0;
                            foreach (var position in Polygons[i].Vertices)
                            {
                                coords[index] = new CLLocationCoordinate2D(latitude: position.Latitude, longitude: position.Longitude);
                                index++;
                            }
                            PolygonIOS blockOverlay = new PolygonIOS(MKPolygon.FromCoordinates(coords));
                            blockOverlay.FillColor = Color.FromUint((uint)Polygons[i].Color).ToUIColor();
                            blockOverlay.StrokeColor = Color.FromUint((uint)Polygons[i].StrokeColor).ToUIColor();
                            nativeMap.AddOverlay(blockOverlay);
                        }
                    }
                    catch (Exception ex)
                    {
                        Debug.WriteLine(ex);
                    }
                }

                UITapGestureRecognizer tap = new UITapGestureRecognizer(g =>
                {
                    var pt = g.LocationInView(nativeMap);
                    CLLocationCoordinate2D tapCoord = nativeMap.ConvertPoint(pt, nativeMap);
                    Console.WriteLine("new CLLocationCoordinate2D({0}, {1}),", tapCoord.Latitude, tapCoord.Longitude);
                    Position clickedPos = new Position(tapCoord.Latitude, tapCoord.Longitude);
                    formsMap.OniOSMapClicked(this, clickedPos, e); //this method verify if touch was inside or outside of the bounds of the polygon using Latitude and Longitude
                    RefreshPolygons(formsMap); //render again and update colors
                });
                nativeMap.AddGestureRecognizer(tap);


                MKOverlayRenderer GetOverlayRenderer(MKMapView mapView, IMKOverlay overlayWrapper)
                {
                    return new MKPolygonRenderer(overlayWrapper as MKPolygon)
                    {
                        FillColor = (overlayWrapper as PolygonIOS).FillColor,
                        StrokeColor = (overlayWrapper as PolygonIOS).StrokeColor,
                        Alpha = 0.4f,
                        LineWidth = 2
                    };
                }
            }
```

In my UITapGestureRecognizer I have one RefreshPolygons, that is one method just to render again the map and update polygon color. In this way user can select and deselect sectors and visualize this on map.

But my problem is: After the user select one or more polygons, the ok button is expected to be touched, in this moment another page will open, but when I touch the button I have an application output message:

```
objc[2042]: Attempt to use unknown class 0x3cee07ae7870.

=================================================================
	Native Crash Reporting
=================================================================
Got a SIGABRT while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries 
used by your application.
=================================================================

=================================================================
	Basic Fault Adddress Reporting
=================================================================
Memory around native instruction pointer (0x117141eb6):0x117141ea6  ff ff c3 90 90 90 b8 09 02 00 02 49 89 ca 0f 05  ...........I....
0x117141eb6  73 08 48 89 c7 e9 29 7c ff ff c3 90 90 90 b8 1e  s.H...)|........
0x117141ec6  00 00 02 49 89 ca 0f 05 73 08 48 89 c7 e9 36 7c  ...I....s.H...6|
0x117141ed6  ff ff c3 90 90 90 b8 94 01 00 02 49 89 ca 0f 05  ...........I....

=================================================================
	Native stacktrace:
=================================================================
	0x10793e960 - /Users/matheuscrespischenfeld/Library/Developer/CoreSimulator/Devices/A1EADC21-E4A6-4A0D-BFBB-23326D7380BE/data/Containers/Bundle/Application/839956A2-4B96-400F-B257-F89BAB78029E/Elysios.Demetra.App.iOS.app/Frameworks/Mono.framework/Mono : mono_dump_native_crash_info
	0x1078e80c5 - /Users/matheuscrespischenfeld/Library/Developer/CoreSimulator/Devices/A1EADC21-E4A6-4A0D-BFBB-23326D7380BE/data/Containers/Bundle/Application/839956A2-4B96-400F-B257-F89BAB78029E/Elysios.Demetra.App.iOS.app/Frameworks/Mono.framework/Mono : mono_handle_native_crash
	0x117167f5a - /usr/lib/system/libsystem_platform.dylib : _sigtramp
	0x7000029151e0 - Unknown
	0x11713c25f - /usr/lib/system/libsystem_kernel.dylib : abort_with_payload_wrapper_internal
	0x115f4f1f5 - /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/Library/CoreSimulator/Profiles/Runtimes/iOS.simruntime/Contents/Resources/RuntimeRoot/usr/lib/libobjc.A.dylib : _ZL12_objc_fatalvyyPKcP13__va_list_tag
	0x115f4f11e - /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/Library/CoreSimulator/Profiles/Runtimes/iOS.simruntime/Contents/Resources/RuntimeRoot/usr/lib/libobjc.A.dylib : __objc_error
	0x115f56c18 - /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/Library/CoreSimulator/Profiles/Runtimes/iOS.simruntime/Contents/Resources/RuntimeRoot/usr/lib/libobjc.A.dylib : class_copyProtocolList
	0x10747aa53 - /Users/matheuscrespischenfeld/Library/Developer/CoreSimulator/Devices/A1EADC21-E4A6-4A0D-BFBB-23326D7380BE/data/Containers/Bundle/Application/839956A2-4B96-400F-B257-F89BAB78029E/Elysios.Demetra.App.iOS.app/Elysios.Demetra.App.iOS : _ZL22get_method_descriptionP10objc_classP13objc_selector
	0x10747a6cb - /Users/matheuscrespischenfeld/Library/Developer/CoreSimulator/Devices/A1EADC21-E4A6-4A0D-BFBB-23326D7380BE/data/Containers/Bundle/Application/839956A2-4B96-400F-B257-F89BAB78029E/Elysios.Demetra.App.iOS.app/Elysios.Demetra.App.iOS : xamarin_get_frame_length
	0x10748967c - /Users/matheuscrespischenfeld/Library/Developer/CoreSimulator/Devices/A1EADC21-E4A6-4A0D-BFBB-23326D7380BE/data/Containers/Bundle/Application/839956A2-4B96-400F-B257-F89BAB78029E/Elysios.Demetra.App.iOS.app/Elysios.Demetra.App.iOS : xamarin_dyn_objc_msgSend
	0x13b2cec80 - Unknown
	0x13af3bd7a - Unknown
	0x107aa014f - /Users/matheuscrespischenfeld/Library/Developer/CoreSimulator/Devices/A1EADC21-E4A6-4A0D-BFBB-23326D7380BE/data/Containers/Bundle/Application/839956A2-4B96-400F-B257-F89BAB78029E/Elysios.Demetra.App.iOS.app/Frameworks/Mono.framework/Mono : mono_gc_run_finalize
	0x107abfbda - /Users/matheuscrespischenfeld/Library/Developer/CoreSimulator/Devices/A1EADC21-E4A6-4A0D-BFBB-23326D7380BE/data/Containers/Bundle/Application/839956A2-4B96-400F-B257-F89BAB78029E/Elysios.Demetra.App.iOS.app/Frameworks/Mono.framework/Mono : sgen_gc_invoke_finalizers
	0x107aa1f03 - /Users/matheuscrespischenfeld/Library/Developer/CoreSimulator/Devices/A1EADC21-E4A6-4A0D-BFBB-23326D7380BE/data/Containers/Bundle/Application/839956A2-4B96-400F-B257-F89BAB78029E/Elysios.Demetra.App.iOS.app/Frameworks/Mono.framework/Mono : finalizer_thread
	0x107a4f409 - /Users/matheuscrespischenfeld/Library/Developer/CoreSimulator/Devices/A1EADC21-E4A6-4A0D-BFBB-23326D7380BE/data/Containers/Bundle/Application/839956A2-4B96-400F-B257-F89BAB78029E/Elysios.Demetra.App.iOS.app/Frameworks/Mono.framework/Mono : start_wrapper_internal
	0x107a4f267 - /Users/matheuscrespischenfeld/Library/Developer/CoreSimulator/Devices/A1EADC21-E4A6-4A0D-BFBB-23326D7380BE/data/Containers/Bundle/Application/839956A2-4B96-400F-B257-F89BAB78029E/Elysios.Demetra.App.iOS.app/Frameworks/Mono.framework/Mono : start_wrapper
	0x117179661 - /usr/lib/system/libsystem_pthread.dylib : _pthread_body
	0x11717950d - /usr/lib/system/libsystem_pthread.dylib : _pthread_body
	0x117178bf9 - /usr/lib/system/libsystem_pthread.dylib : thread_start

=================================================================
	Managed Stacktrace:
=================================================================
	  at <unknown> <0xffffffff>
	  at ObjCRuntime.Messaging:void_objc_msgSend <0x000bf>
	  at ObjCRuntime.BaseWrapper:Dispose <0x00162>
	  at ObjCRuntime.BaseWrapper:Finalize <0x0006b>
	  at System.Object:runtime_invoke_virtual_void__this__ <0x000c9>
=================================================================
```

So it makes me think that my problem is in my `e.OldElement` that is:

```
if (e.OldElement != null)
            {
                var nativeMap = Control as MKMapView;
                //var formsMap = (CustomMap)e.OldElement;
                if (nativeMap != null)
                {
                    nativeMap.RemoveOverlays(nativeMap.Overlays);
                    nativeMap.OverlayRenderer = null;
                    //nativeMap.RemoveGestureRecognizer(nativeMap.GestureRecognizers[nativeMap.GestureRecognizers.Length-1]);
                    //nativeMap.GetViewForOverlay = null;
                }
            }
```
I already tried to isolate with try catch, tried to use break points and identify exactly where the error occurs, tried to use different codes in e.OldElement but the error still happening. 

I am available to explain the situation better or help with any questions, but I really need help to resolve this.

**Comment by hartez** [comment]
@matheuscschenfeld You should open a separate GitHub issue for this problem.

**Pull request merged**